From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uazo <uazo@users.noreply.github.com>
Date: Wed, 21 Sep 2022 12:28:17 +0000
Subject: Remove HTTP referrals in cross origin navigation

The patch removes the referrals if the navigation is cross-origin and occurs in the top frame.
The value is not removed across iframes because the referrals are statically defined
by the HTML and the Javascript of the page, and therefore they do not tell anything about the
user. Also, some services may not work, such as video iframes.
A preference is also introduced to completely removes referrals management, for advanced users.

Original License: GPL-2.0-or-later - https://spdx.org/licenses/GPL-2.0-or-later.html
License: GPL-3.0-only - https://spdx.org/licenses/GPL-3.0-only.html
---
 chrome/android/chrome_java_resources.gni      |  2 +
 chrome/android/chrome_java_sources.gni        |  2 +
 ...button_group_referer_policy_preference.xml | 47 +++++++++
 .../java/res/xml/privacy_preferences.xml      |  4 +
 .../res/xml/referer_policy_preferences.xml    | 31 ++++++
 .../privacy/settings/PrivacySettings.java     |  6 ++
 .../RadioButtonGroupRefererSettings.java      | 89 +++++++++++++++++
 .../settings/RefererSettingsFragment.java     | 98 +++++++++++++++++++
 .../api/settings_private/prefs_util.cc        |  5 +
 .../net/system_network_context_manager.cc     |  4 +
 chrome/browser/renderer_preferences_util.cc   |  1 +
 .../settings/privacy_page/privacy_page.html   | 10 ++
 .../settings/privacy_page/privacy_page.ts     | 18 ++++
 .../strings/android_chrome_strings.grd        | 24 +++++
 chrome/browser/ui/prefs/pref_watcher.cc       |  2 +
 chrome/browser/ui/prefs/prefs_tab_helper.cc   |  2 +
 .../settings_localized_strings_provider.cc    |  4 +
 chrome/common/pref_names.h                    |  3 +
 ...-referrals-in-cross-origin-navigation.grdp | 18 ++++
 .../content/internal/context_menu_download.cc | 10 ++
 .../renderer_host/navigation_request.cc       |  9 ++
 .../network/public/cpp/resource_request.h     |  2 +-
 .../renderer_preferences.h                    |  1 +
 23 files changed, 391 insertions(+), 1 deletion(-)
 create mode 100644 chrome/android/java/res/layout/radio_button_group_referer_policy_preference.xml
 create mode 100644 chrome/android/java/res/xml/referer_policy_preferences.xml
 create mode 100644 chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RadioButtonGroupRefererSettings.java
 create mode 100644 chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RefererSettingsFragment.java
 create mode 100644 components/cromite_components_strings_grd/Remove-HTTP-referrals-in-cross-origin-navigation.grdp

diff --git a/chrome/android/chrome_java_resources.gni b/chrome/android/chrome_java_resources.gni
index 7ae73a62b3153f45ca3b77e9badce8c8a4485325..3860d579a5a5a2dd9be00904246b32c8b03e82e5 100644
--- a/chrome/android/chrome_java_resources.gni
+++ b/chrome/android/chrome_java_resources.gni
@@ -690,6 +690,8 @@ chrome_java_resources = [
   "java/res/xml/personalize_google_services_preferences.xml",
   "java/res/xml/incognito_preferences.xml",
   "java/res/xml/privacy_preferences.xml",
+  "java/res/xml/referer_policy_preferences.xml",
+  "java/res/layout/radio_button_group_referer_policy_preference.xml",
   "java/res/xml/search_widget_info.xml",
   "java/res/xml/tracing_preferences.xml",
   "java/res/xml/unified_account_settings_preferences.xml",
diff --git a/chrome/android/chrome_java_sources.gni b/chrome/android/chrome_java_sources.gni
index 8c21719083926464a9aaa6ccea757a17ccaebc41..137019df16f21e17af9f48a2b82c8f47e5354a58 100644
--- a/chrome/android/chrome_java_sources.gni
+++ b/chrome/android/chrome_java_sources.gni
@@ -1051,6 +1051,8 @@ chrome_java_sources = [
   "java/src/org/chromium/chrome/browser/privacy/settings/UniversalOptOutSettings.java",
   "java/src/org/chromium/chrome/browser/privacy/settings/IncognitoSettings.java",
   "java/src/org/chromium/chrome/browser/privacy_sandbox/CctHandler.java",
+  "java/src/org/chromium/chrome/browser/privacy/settings/RefererSettingsFragment.java",
+  "java/src/org/chromium/chrome/browser/privacy/settings/RadioButtonGroupRefererSettings.java",
   "java/src/org/chromium/chrome/browser/provider/BaseColumns.java",
   "java/src/org/chromium/chrome/browser/provider/BookmarkColumns.java",
   "java/src/org/chromium/chrome/browser/provider/ChromeBrowserProviderImpl.java",
diff --git a/chrome/android/java/res/layout/radio_button_group_referer_policy_preference.xml b/chrome/android/java/res/layout/radio_button_group_referer_policy_preference.xml
new file mode 100644
index 0000000000000000000000000000000000000000..22c81fd6f2fd656756a6775d39a2cb66c4a384d1
--- /dev/null
+++ b/chrome/android/java/res/layout/radio_button_group_referer_policy_preference.xml
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    This file is part of Bromite.
+
+    Bromite is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Bromite is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
+-->
+
+<org.chromium.components.browser_ui.widget.RadioButtonWithDescriptionLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:focusable="false">
+
+    <org.chromium.components.browser_ui.widget.RadioButtonWithDescription
+        android:id="@+id/no_cross_origin_top_frame_policy"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:primaryText="@string/no_cross_origin_top_frame_policy_title"
+        app:descriptionText="@string/no_cross_origin_top_frame_policy_summary" />
+
+    <org.chromium.components.browser_ui.widget.RadioButtonWithDescription
+        android:id="@+id/standard_referer_policy"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:primaryText="@string/standard_referer_policy_title"
+        app:descriptionText="@string/standard_referer_policy_summary" />
+
+    <org.chromium.components.browser_ui.widget.RadioButtonWithDescription
+        android:id="@+id/disable_referer_policy"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:primaryText="@string/disable_referer_policy_title"
+        app:descriptionText="@string/disable_referer_policy_summary" />
+
+</org.chromium.components.browser_ui.widget.RadioButtonWithDescriptionLayout>
diff --git a/chrome/android/java/res/xml/privacy_preferences.xml b/chrome/android/java/res/xml/privacy_preferences.xml
index 9e7b9f49e90000e1866de6f009846ab62176350a..ddde72a64950834382fd3320b55216c03ebf19e6 100644
--- a/chrome/android/java/res/xml/privacy_preferences.xml
+++ b/chrome/android/java/res/xml/privacy_preferences.xml
@@ -112,6 +112,10 @@ found in the LICENSE file.
         android:title="@string/close_tabs_on_exit_title"
         android:summary="@string/close_tabs_on_exit_summary"
         android:defaultValue="false" />
+    <Preference
+        android:key="referers_policy"
+        android:title="@string/referers_policy_title"
+        android:fragment="org.chromium.chrome.browser.privacy.settings.RefererSettingsFragment"/>
     <PreferenceCategory
         android:key="services_category"
         android:title="@string/services_category_title">
diff --git a/chrome/android/java/res/xml/referer_policy_preferences.xml b/chrome/android/java/res/xml/referer_policy_preferences.xml
new file mode 100644
index 0000000000000000000000000000000000000000..6788bfe9c39c2b2a1ea74f24eb2df303dd91ba0f
--- /dev/null
+++ b/chrome/android/java/res/xml/referer_policy_preferences.xml
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    This file is part of Bromite.
+
+    Bromite is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Bromite is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
+-->
+
+<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <org.chromium.components.browser_ui.settings.TextMessagePreference
+        android:key="summary"
+        android:summary="@string/referers_policy_summary"
+        app:allowDividerBelow="false" />
+    <org.chromium.chrome.browser.privacy.settings.RadioButtonGroupRefererSettings
+        android:key="referer_page_radio_button_group"
+        android:selectable="false"
+        app:allowDividerAbove="false"
+        app:allowDividerBelow="false" />
+</PreferenceScreen>
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
index 041b6c6fc16130a2167df3c7e838f1faa3a34484..efe7da82e85e386dbc38daafa75b4fad8e0fc04a 100644
--- a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/PrivacySettings.java
@@ -44,6 +44,7 @@ import org.chromium.chrome.browser.preferences.ChromePreferenceKeys;
 import org.chromium.chrome.browser.preferences.ChromeSharedPreferences;
 import org.chromium.chrome.browser.preferences.Pref;
 import org.chromium.chrome.browser.prefetch.settings.PreloadPagesSettingsFragment;
+import org.chromium.chrome.browser.privacy.settings.RefererSettingsFragment;
 import org.chromium.chrome.browser.privacy.secure_dns.SecureDnsSettings;
 import org.chromium.chrome.browser.privacy_guide.PrivacyGuideFragment;
 import org.chromium.chrome.browser.privacy_guide.PrivacyGuideInteractions;
@@ -119,6 +120,7 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
     private static final String PREF_SAFE_BROWSING = "safe_browsing";
     private static final String PREF_PASSWORD_LEAK_DETECTION = "password_leak_detection";
     private static final String PREF_SYNC_AND_SERVICES_LINK = "sync_and_services_link";
+    private static final String PREF_REFERER_POLICY = "referers_policy";
     private static final String PREF_PRIVACY_GUIDE = "privacy_guide";
     private static final String PREF_PROXY_OPTIONS = "proxy";
     private static final String PREF_INCOGNITO_LOCK = "incognito_lock";
@@ -538,6 +540,10 @@ public class PrivacySettings extends ChromeBaseSettingsFragment
             }
         }
 
+        Preference refererPolicyPref = findPreference(PREF_REFERER_POLICY);
+        refererPolicyPref.setSummary(
+                RefererSettingsFragment.getRefererSummaryString(getContext()));
+
         Preference preloadPagesPreference = findPreference(PREF_PRELOAD_PAGES);
         if (preloadPagesPreference != null) {
             preloadPagesPreference.setSummary(
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RadioButtonGroupRefererSettings.java b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RadioButtonGroupRefererSettings.java
new file mode 100644
index 0000000000000000000000000000000000000000..a9d2f0c5d77db76c61d5b9adcbefab460e0e00a3
--- /dev/null
+++ b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RadioButtonGroupRefererSettings.java
@@ -0,0 +1,89 @@
+/*
+    This file is part of Bromite.
+
+    Bromite is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Bromite is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package org.chromium.chrome.browser.privacy.settings;
+
+import android.content.Context;
+import android.util.AttributeSet;
+import android.view.View;
+import android.widget.RadioGroup;
+
+import androidx.preference.Preference;
+import androidx.preference.PreferenceViewHolder;
+
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.flags.ChromeFeatureList;
+import org.chromium.components.browser_ui.widget.RadioButtonWithDescription;
+import org.chromium.components.browser_ui.widget.RadioButtonWithDescriptionLayout;
+
+public class RadioButtonGroupRefererSettings extends Preference
+        implements RadioGroup.OnCheckedChangeListener {
+
+    private RadioButtonWithDescription mStandardReferer;
+    private RadioButtonWithDescription mNoCrossOriginReferer;
+    private RadioButtonWithDescription mNoReferer;
+    private int mRefererPolicy;
+
+    public RadioButtonGroupRefererSettings(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        setLayoutResource(R.layout.radio_button_group_referer_policy_preference);
+    }
+
+    public void init(boolean enableReferer, int refererPolicy) {
+        if (!enableReferer)
+            mRefererPolicy = 0;
+        else
+            mRefererPolicy = refererPolicy;
+    }
+
+    @Override
+    public void onCheckedChanged(RadioGroup group, int checkedId) {
+        if (checkedId == mStandardReferer.getId()) {
+            mRefererPolicy = 1;
+        } else if (checkedId == mNoCrossOriginReferer.getId()) {
+            mRefererPolicy = 2;
+        } else if (checkedId == mNoReferer.getId()) {
+            mRefererPolicy = 0;
+        } else {
+            assert false : "Should not be reached.";
+        }
+        callChangeListener(mRefererPolicy);
+    }
+
+    @Override
+    public void onBindViewHolder(PreferenceViewHolder holder) {
+        super.onBindViewHolder(holder);
+
+        mStandardReferer = (RadioButtonWithDescription) holder.findViewById(
+                R.id.standard_referer_policy);
+        mNoCrossOriginReferer = (RadioButtonWithDescription) holder.findViewById(
+                R.id.no_cross_origin_top_frame_policy);
+        mNoReferer = (RadioButtonWithDescription) holder.findViewById(R.id.disable_referer_policy);
+
+        RadioButtonWithDescriptionLayout groupLayout =
+                (RadioButtonWithDescriptionLayout) mNoReferer.getRootView();
+        setCheckedState(mRefererPolicy);
+        groupLayout.setOnCheckedChangeListener(this);
+    }
+
+    public void setCheckedState(int checkedState) {
+        mRefererPolicy = checkedState;
+        mStandardReferer.setChecked(checkedState == 1);
+        mNoCrossOriginReferer.setChecked(checkedState == 2);
+        mNoReferer.setChecked(checkedState == 0);
+    }
+}
diff --git a/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RefererSettingsFragment.java b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RefererSettingsFragment.java
new file mode 100644
index 0000000000000000000000000000000000000000..23bb3140ee6865f470353614dbdaa11f07e2c8ce
--- /dev/null
+++ b/chrome/android/java/src/org/chromium/chrome/browser/privacy/settings/RefererSettingsFragment.java
@@ -0,0 +1,98 @@
+/*
+    This file is part of Bromite.
+
+    Bromite is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Bromite is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Bromite. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package org.chromium.chrome.browser.privacy.settings;
+
+import android.os.Bundle;
+import android.content.Context;
+
+import androidx.preference.Preference;
+import androidx.preference.PreferenceFragmentCompat;
+
+import org.chromium.base.supplier.MonotonicObservableSupplier;
+import org.chromium.base.supplier.ObservableSuppliers;
+import org.chromium.base.supplier.SettableMonotonicObservableSupplier;
+
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.preferences.Pref;
+import org.chromium.chrome.browser.profiles.Profile;
+import org.chromium.chrome.browser.profiles.ProfileManager;
+import org.chromium.chrome.browser.settings.ChromeBaseSettingsFragment;
+import org.chromium.components.browser_ui.settings.SettingsFragment;
+import org.chromium.components.browser_ui.settings.SettingsUtils;
+import org.chromium.components.browser_ui.settings.TextMessagePreference;
+import org.chromium.components.prefs.PrefService;
+import org.chromium.components.user_prefs.UserPrefs;
+
+public class RefererSettingsFragment extends ChromeBaseSettingsFragment
+        implements Preference.OnPreferenceChangeListener {
+
+    static final String PREF_REFERER_POLICY = "referer_page_radio_button_group";
+
+    private RadioButtonGroupRefererSettings mRefererPreference;
+    private final PrefService prefService = UserPrefs.get(ProfileManager.getLastUsedRegularProfile());
+
+    private final SettableMonotonicObservableSupplier<String> mPageTitle =
+        ObservableSuppliers.createMonotonic();
+
+    public static String getRefererSummaryString(Context context) {
+        int policy = UserPrefs.get(ProfileManager.getLastUsedRegularProfile())
+                              .getInteger(Pref.REFERRERS_POLICY);
+
+        if (policy == 0) {
+            return context.getString(R.string.disable_referer_policy_title);
+        }
+        if (policy == 1) {
+            return context.getString(R.string.standard_referer_policy_title);
+        }
+        if (policy == 2) {
+            return context.getString(R.string.no_cross_origin_top_frame_policy_title);
+        }
+        assert false : "Should not be reached";
+        return "";
+    }
+
+    @Override
+    public @SettingsFragment.AnimationType int getAnimationType() {
+        return SettingsFragment.AnimationType.PROPERTY;
+    }
+
+    @Override
+    public MonotonicObservableSupplier<String> getPageTitle() {
+        return mPageTitle;
+    }
+
+    @Override
+    public void onCreatePreferences(Bundle bundle, String s) {
+        SettingsUtils.addPreferencesFromResource(this, R.xml.referer_policy_preferences);
+        getActivity().setTitle(R.string.referers_policy_title);
+
+        mRefererPreference = findPreference(PREF_REFERER_POLICY);
+        mRefererPreference.init(prefService.getBoolean(Pref.ENABLE_REFERRERS),
+                                prefService.getInteger(Pref.REFERRERS_POLICY));
+        mRefererPreference.setOnPreferenceChangeListener(this);
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        UserPrefs.get(ProfileManager.getLastUsedRegularProfile())
+                 .setInteger(Pref.REFERRERS_POLICY, (int)newValue);
+        UserPrefs.get(ProfileManager.getLastUsedRegularProfile())
+                 .setBoolean(Pref.ENABLE_REFERRERS, (int)newValue != 0);
+        return true;
+    }
+}
diff --git a/chrome/browser/extensions/api/settings_private/prefs_util.cc b/chrome/browser/extensions/api/settings_private/prefs_util.cc
index a3c5088bf899d5f49b1a356aff6fd7ed8befe350..842c0c28963a2b09936733087c8f93d91a56de96 100644
--- a/chrome/browser/extensions/api/settings_private/prefs_util.cc
+++ b/chrome/browser/extensions/api/settings_private/prefs_util.cc
@@ -319,6 +319,11 @@ const PrefsUtil::TypedPrefMap& PrefsUtil::GetAllowlistedKeys() {
   (*s_allowlist)[::prefs::kExpireDaysThreshold] =
       settings_api::PrefType::kNumber;
 
+  (*s_allowlist)[::prefs::kEnableReferrers] =
+      settings_api::PrefType::kBoolean;
+  (*s_allowlist)[::prefs::kReferrersPolicy] =
+      settings_api::PrefType::kNumber;
+
   // Appearance settings.
   (*s_allowlist)[::prefs::kCtrlTabMru] = settings_api::PrefType::kBoolean;
   (*s_allowlist)[::prefs::kCurrentThemeID] = settings_api::PrefType::kString;
diff --git a/chrome/browser/net/system_network_context_manager.cc b/chrome/browser/net/system_network_context_manager.cc
index 90f2b0cd768ac2f8c022b41e950612fb22d7188a..8e79e54f5d702876a90182bb32a49b76f7b9b79a 100644
--- a/chrome/browser/net/system_network_context_manager.cc
+++ b/chrome/browser/net/system_network_context_manager.cc
@@ -676,6 +676,9 @@ SystemNetworkContextManager::SystemNetworkContextManager(
       base::BindRepeating(&SystemNetworkContextManager::UpdateReferrersEnabled,
                           base::Unretained(this)));
 
+  local_state_->SetDefaultPrefValue(
+      prefs::kReferrersPolicy, base::Value(2));
+
   pref_change_registrar_.Add(
       prefs::kExplicitlyAllowedNetworkPorts,
       base::BindRepeating(
@@ -758,6 +761,7 @@ void SystemNetworkContextManager::RegisterPrefs(PrefRegistrySimple* registry) {
   // the system NetworkContext, and the per-profile pref values are used for
   // the profile NetworkContexts.
   registry->RegisterBooleanPref(prefs::kEnableReferrers, true);
+  registry->RegisterIntegerPref(prefs::kReferrersPolicy, 2);
 
   registry->RegisterBooleanPref(prefs::kQuickCheckEnabled, true);
 
diff --git a/chrome/browser/renderer_preferences_util.cc b/chrome/browser/renderer_preferences_util.cc
index 2b542a4814ddc390ce3d9d093a61e83188889352..b6868f80c14e6312ee3a96a58d3c79c7f4b5887f 100644
--- a/chrome/browser/renderer_preferences_util.cc
+++ b/chrome/browser/renderer_preferences_util.cc
@@ -153,6 +153,7 @@ void UpdateFromSystemSettings(blink::RendererPreferences* prefs,
   prefs->caret_blink_interval =
       ui::NativeTheme::GetInstanceForNativeUi()->caret_blink_interval();
   prefs->enable_referrers = pref_service->GetBoolean(prefs::kEnableReferrers);
+  prefs->referrers_policy = pref_service->GetInteger(prefs::kReferrersPolicy);
   prefs->enable_do_not_track =
       pref_service->GetBoolean(prefs::kEnableDoNotTrack);
   // TODO(crbug.com/40745270): Clean up kGlobalPrivacyControlForce &
diff --git a/chrome/browser/resources/settings/privacy_page/privacy_page.html b/chrome/browser/resources/settings/privacy_page/privacy_page.html
index c7c47e7a9c63c5bda04593c9027a93fff50575c7..f99c6f4600bbfa24906c2740352c05c7e5177caf 100644
--- a/chrome/browser/resources/settings/privacy_page/privacy_page.html
+++ b/chrome/browser/resources/settings/privacy_page/privacy_page.html
@@ -52,6 +52,16 @@
           menu-options="[[historyExpireDaysThresholdOptions_]]">
       </settings-dropdown-menu>
     </div>
+    <div class="cr-row">
+      <div class="flex cr-padded-text" aria-hidden="true">
+        $i18n{referersPolicyTitle}
+      </div>
+      <settings-dropdown-menu id="referersPolicy" label="$i18n{referersPolicyTitle}"
+          pref="{{prefs.referrers_policy}}"
+          menu-options="[[referersPolicyOptions_]]"
+          on-settings-control-change="onReferersPolicyChanged_">
+      </settings-dropdown-menu>
+    </div>
     <cr-toast id="deleteBrowsingDataToast" duration="5000">
       <div>[[dbdDeletionConfirmationToastLabel_]]</div>
     </cr-toast>
diff --git a/chrome/browser/resources/settings/privacy_page/privacy_page.ts b/chrome/browser/resources/settings/privacy_page/privacy_page.ts
index 8cb0cc583dd6a8d6b6da82442d49098b6a364cc4..b825d503efa12fab6c69f5014fe6739c7210e15c 100644
--- a/chrome/browser/resources/settings/privacy_page/privacy_page.ts
+++ b/chrome/browser/resources/settings/privacy_page/privacy_page.ts
@@ -84,6 +84,17 @@ export class SettingsPrivacyPageElement extends SettingsPrivacyPageElementBase {
       showClearBrowsingDataDialog_: Boolean,
       showPrivacyGuideDialog_: Boolean,
 
+      referersPolicyOptions_: {
+        readOnly: true,
+        type: Array,
+        value() {
+          return [
+            {value: 2, name: loadTimeData.getString('noCrossOriginTopFramePolicy')},
+            {value: 1, name: loadTimeData.getString('standardRefererPolicy')},
+            {value: 0, name: loadTimeData.getString('disableRefererPolicy')},
+          ];
+        },
+      },
       // The label of the confirmation toast that is displayed after deletion
       // from 'Delete Browsing data' is completed.
       dbdDeletionConfirmationToastLabel_: {
@@ -119,6 +130,7 @@ export class SettingsPrivacyPageElement extends SettingsPrivacyPageElementBase {
   declare private historyExpireDaysThresholdOptions_: object[];
   declare private showClearBrowsingDataDialog_: boolean;
   declare private showPrivacyGuideDialog_: boolean;
+  declare private referersPolicyOptions_: object[];
   declare private dbdDeletionConfirmationToastLabel_: string;
   declare private shouldShowDbdDeletionConfirmationToast_: boolean;
   declare private showUniversalOptOutSettings_: boolean;
@@ -150,6 +162,12 @@ export class SettingsPrivacyPageElement extends SettingsPrivacyPageElementBase {
     Router.getInstance().navigateTo(routes.COOKIES);
   }
 
+  private onReferersPolicyChanged_(): void {
+    const policy =
+        this.getPref('referrers_policy').value;
+    this.setPrefValue('enable_referrers', policy != 0);
+  }
+
   private onCbdDialogClosed_() {
     Router.getInstance().navigateTo(routes.CLEAR_BROWSER_DATA.parent!);
 
diff --git a/chrome/browser/ui/android/strings/android_chrome_strings.grd b/chrome/browser/ui/android/strings/android_chrome_strings.grd
index 1c3d70b30dd353999ca53eb2b57b94f416f77374..8d889caa1ac2a8e9df4f5edd7bafa9e54859170b 100644
--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
+++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
@@ -1254,6 +1254,30 @@ For example, some websites may respond to this request by showing you ads that a
       <message name="IDS_UNIVERSAL_OPT_OUT_INFO_TEXT" desc="Subtitle of a setting. It appears below the title: ‘Send an opt-out request to sites regarding use of your data’. Users choose whether they want to send a request to sites about how to treat their user data. This string explains that sites may not respond to the user’s request even when this toggle is on; and, even when they do respond, sites may handle opt-out requests differently based on regional laws. This string appears on its own subpage. The tone should be informative. Clicking on Learn more link opens a help center article, the link tag should not be translated but the 'learn more' text inside the link tag should be translated.">
         The sites you visit, including Google, are responsible for determining how they respond to this request. Depending on the laws in your region, sites may respond by limiting the sale and sharing of your data or its use for personalized ads. <ph name="BEGIN_LINK">&lt;link&gt;</ph>Learn more<ph name="END_LINK">&lt;/link&gt;</ph>
       </message>
+      <message name="IDS_REFERERS_POLICY_TITLE" desc="Title for 'Enable referers' preference">
+        HTTP Referer header policy
+      </message>
+      <message name="IDS_REFERERS_POLICY_SUMMARY" desc="Summary for a section in Settings that controls referer policy.">
+        Allow to change default HTTP Referer header policy
+      </message>
+      <message name="IDS_STANDARD_REFERER_POLICY_TITLE" desc="Title for 'standard referer' preference">
+        Standard HTTP Referrer header policy
+      </message>
+      <message name="IDS_STANDARD_REFERER_POLICY_SUMMARY" desc="Summary for 'standard referer' preference">
+        Send only origin if the request is cross-origin
+      </message>
+      <message name="IDS_NO_CROSS_ORIGIN_TOP_FRAME_POLICY_TITLE" desc="Title for 'standard referer' preference">
+        Disable if cross-origin and user begins navigation
+      </message>
+      <message name="IDS_NO_CROSS_ORIGIN_TOP_FRAME_POLICY_SUMMARY" desc="Summary for 'no cross origin referer' preference">
+        Remove referrer if the navigation is cross-origin and occurs in the top frame otherwise send only origin if the request is cross-origin
+      </message>
+      <message name="IDS_DISABLE_REFERER_POLICY_TITLE" desc="Title for 'disable referer' preference">
+        Disable all referrers
+      </message>
+      <message name="IDS_DISABLE_REFERER_POLICY_SUMMARY" desc="Summary for 'disable referer' preference">
+        Disable all referrers, may break navigation
+      </message>
       <message name="IDS_CAN_MAKE_PAYMENT_TITLE" desc="Title for preference to allow websites to know whether you have payment methods available through PaymentRequest.CanMakePayment interface">
         Access payment methods
       </message>
diff --git a/chrome/browser/ui/prefs/pref_watcher.cc b/chrome/browser/ui/prefs/pref_watcher.cc
index 7b02f5e7a3649f0f5f7490961c1477eadc423fb2..70cd74d51457fc0322a07115701cb732649bc83e 100644
--- a/chrome/browser/ui/prefs/pref_watcher.cc
+++ b/chrome/browser/ui/prefs/pref_watcher.cc
@@ -137,6 +137,8 @@ PrefWatcher::PrefWatcher(Profile* profile) : profile_(profile) {
                                      renderer_callback);
   profile_pref_change_registrar_.Add(prefs::kEnableReferrers,
                                      renderer_callback);
+  profile_pref_change_registrar_.Add(prefs::kReferrersPolicy,
+                                     renderer_callback);
   profile_pref_change_registrar_.Add(prefs::kEnableEncryptedMedia,
                                      renderer_callback);
   profile_pref_change_registrar_.Add(prefs::kWebRTCIPHandlingPolicy,
diff --git a/chrome/browser/ui/prefs/prefs_tab_helper.cc b/chrome/browser/ui/prefs/prefs_tab_helper.cc
index 66eede68f09d2c29f3d5c6997c24b73710b4f4fd..5421cb6e7b4fd736230bc76ce194077d38fb753d 100644
--- a/chrome/browser/ui/prefs/prefs_tab_helper.cc
+++ b/chrome/browser/ui/prefs/prefs_tab_helper.cc
@@ -388,6 +388,8 @@ void PrefsTabHelper::RegisterProfilePrefs(
   registry->RegisterBooleanPref(
       prefs::kEnableReferrers,
       !base::FeatureList::IsEnabled(features::kNoReferrers));
+  registry->RegisterIntegerPref(
+      prefs::kReferrersPolicy, 2); // NoCrossOriginReferer
   registry->RegisterBooleanPref(prefs::kEnableEncryptedMedia, true);
   registry->RegisterBooleanPref(prefs::kScrollToTextFragmentEnabled, false);
 #if BUILDFLAG(IS_ANDROID)
diff --git a/chrome/browser/ui/webui/settings/settings_localized_strings_provider.cc b/chrome/browser/ui/webui/settings/settings_localized_strings_provider.cc
index a6a3589de1407fc7515bce4b25a8e740c6f4d424..a6aa6a32d8f8fa72271fb78140a20ae09a376de7 100644
--- a/chrome/browser/ui/webui/settings/settings_localized_strings_provider.cc
+++ b/chrome/browser/ui/webui/settings/settings_localized_strings_provider.cc
@@ -2647,6 +2647,10 @@ void AddPrivacyStrings(content::WebUIDataSource* html_source,
       {"dayHistory", IDS_SETTINGS_DAY_HISTORY_DESCRIPTION},
       {"daysHistory", IDS_SETTINGS_DAYS_HISTORY_DESCRIPTION},
       {"foreverHistory", IDS_SETTINGS_FOREVER_HISTORY_DESCRIPTION},
+      {"referersPolicyTitle", IDS_REFERERS_POLICY_TITLE},
+      {"standardRefererPolicy", IDS_STANDARD_REFERER_POLICY_TITLE},
+      {"noCrossOriginTopFramePolicy", IDS_NO_CROSS_ORIGIN_TOP_FRAME_POLICY_TITLE},
+      {"disableRefererPolicy", IDS_DISABLE_REFERER_POLICY_TITLE},
 #if BUILDFLAG(GOOGLE_CHROME_BRANDING)
       {"safeBrowsingStandardDescProxy",
        IDS_SETTINGS_SAFEBROWSING_STANDARD_DESC_PROXY},
diff --git a/chrome/common/pref_names.h b/chrome/common/pref_names.h
index eabcb6077aa0d428b57f432533f80637f992799b..6f344115135ca294721c9831e723137edd290688 100644
--- a/chrome/common/pref_names.h
+++ b/chrome/common/pref_names.h
@@ -910,6 +910,9 @@ inline constexpr char kEnableDoNotTrack[] = "enable_do_not_track";
 inline constexpr char kEnableEncryptedMedia[] =
     "webkit.webprefs.encrypted_media_enabled";
 
+// Set referrer policy.
+inline constexpr char kReferrersPolicy[] = "referrers_policy";
+
 // Boolean that specifies whether to import the form data for autofill from the
 // default browser on first run.
 inline constexpr char kImportAutofillFormData[] = "import_autofill_form_data";
diff --git a/components/cromite_components_strings_grd/Remove-HTTP-referrals-in-cross-origin-navigation.grdp b/components/cromite_components_strings_grd/Remove-HTTP-referrals-in-cross-origin-navigation.grdp
new file mode 100644
index 0000000000000000000000000000000000000000..379e4f09d81a197f37e7a3e6068c337cf7864954
--- /dev/null
+++ b/components/cromite_components_strings_grd/Remove-HTTP-referrals-in-cross-origin-navigation.grdp
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<grit-part>
+  <message name="IDS_REFERERS_POLICY_TITLE" desc="Title for 'Enable referers' preference">
+    HTTP Referer header policy
+  </message>
+  <message name="IDS_REFERERS_POLICY_SUMMARY" desc="Summary for a section in Settings that controls referer policy.">
+    Allow to change default HTTP Referer header policy
+  </message>
+  <message name="IDS_STANDARD_REFERER_POLICY_TITLE" desc="Title for 'standard referer' preference">
+    Standard HTTP Referrer header policy
+  </message>
+  <message name="IDS_NO_CROSS_ORIGIN_TOP_FRAME_POLICY_TITLE" desc="Title for 'standard referer' preference">
+    Disable if cross-origin and user begins navigation (default)
+  </message>
+  <message name="IDS_DISABLE_REFERER_POLICY_TITLE" desc="Title for 'disable referer' preference">
+    Disable all referrers
+  </message>
+</grit-part>
diff --git a/components/download/content/internal/context_menu_download.cc b/components/download/content/internal/context_menu_download.cc
index 0f2aa883b83cd8373f47ba4f496694a8a4a7dbc2..304df6b05c42f85568dfac95c871e739be08373b 100644
--- a/components/download/content/internal/context_menu_download.cc
+++ b/components/download/content/internal/context_menu_download.cc
@@ -13,6 +13,8 @@
 #include "content/public/browser/web_contents.h"
 #include "content/public/common/referrer.h"
 #include "net/traffic_annotation/network_traffic_annotation.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/common/renderer_preferences/renderer_preferences.h"
 
 namespace download {
 
@@ -32,6 +34,14 @@ void CreateContextMenuDownload(const GURL& url,
   content::Referrer referrer = content::Referrer::SanitizeForRequest(
       url,
       content::Referrer(referring_url.GetAsReferrer(), params.referrer_policy));
+  blink::RendererPreferences* render_prefs = web_contents->GetMutableRendererPrefs();
+  if (render_prefs->enable_referrers && render_prefs->referrers_policy == 2 &&
+        !net::registry_controlled_domains::SameDomainOrHost(
+          referring_url, url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
+    referrer = content::Referrer::SanitizeForRequest(
+      url,
+      content::Referrer(params.page_url.GetAsReferrer(), params.referrer_policy));
+  }
   dl_params->set_referrer(referrer.url);
   dl_params->set_referrer_policy(
       content::Referrer::ReferrerPolicyForUrlRequest(referrer.policy));
diff --git a/content/browser/renderer_host/navigation_request.cc b/content/browser/renderer_host/navigation_request.cc
index 0dab916a18711ec84d5b87311a1859773d03311e..f60c70941a5f6bab9188a71f5eb28dd49a0a1eec 100644
--- a/content/browser/renderer_host/navigation_request.cc
+++ b/content/browser/renderer_host/navigation_request.cc
@@ -420,6 +420,15 @@ void AddAdditionalRequestHeaders(
         blink::mojom::Referrer(GURL(), network::mojom::ReferrerPolicy::kNever);
   }
 
+  if (render_prefs.enable_referrers && render_prefs.referrers_policy == 2 &&
+        !net::registry_controlled_domains::SameDomainOrHost(
+          referrer->url, url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES) &&
+        frame_tree_node->IsOutermostMainFrame()) {
+    // remove referrer if the navigation is done on the top frame
+    *referrer =
+        blink::mojom::Referrer(GURL(), network::mojom::ReferrerPolicy::kNever);
+  }
+
   // Next, set the HTTP Origin if needed.
   std::optional<std::string> existing_origin =
       headers->GetHeader(net::HttpRequestHeaders::kOrigin);
diff --git a/services/network/public/cpp/resource_request.h b/services/network/public/cpp/resource_request.h
index 684b6e2f5bcfb3b0926bc897057ae2dc59b69e1f..8c31c71be4931dc0b069e0fa261c71c413500a6a 100644
--- a/services/network/public/cpp/resource_request.h
+++ b/services/network/public/cpp/resource_request.h
@@ -215,7 +215,7 @@ struct COMPONENT_EXPORT(NETWORK_CPP_BASE) ResourceRequest {
   std::vector<GURL> navigation_redirect_chain;
 
   GURL referrer;
-  net::ReferrerPolicy referrer_policy = net::ReferrerPolicy::NEVER_CLEAR;
+  net::ReferrerPolicy referrer_policy = net::ReferrerPolicy::REDUCE_GRANULARITY_ON_TRANSITION_CROSS_ORIGIN;
   net::HttpRequestHeaders headers;
   net::HttpRequestHeaders cors_exempt_headers;
   int load_flags = 0;
diff --git a/third_party/blink/public/common/renderer_preferences/renderer_preferences.h b/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
index 336c8061e4251dc355f9480527d03d46506b61c2..35d41e0cc492e2507020f2ff30860c743e85cffa 100644
--- a/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
+++ b/third_party/blink/public/common/renderer_preferences/renderer_preferences.h
@@ -51,6 +51,7 @@ struct BLINK_COMMON_EXPORT RendererPreferences {
   std::optional<base::TimeDelta> caret_blink_interval;
   bool use_custom_colors{true};
   bool enable_referrers{true};
+  uint32_t referrers_policy{2};
   bool allow_cross_origin_auth_prompt{false};
   bool enable_do_not_track{false};
   bool enable_encrypted_media{true};
